File: packages/pvm/src/host-functions/accumulate/yield.ts
import { bytesToHex, logger } from '@pbnjam/core'
import type { HostFunctionResult } from '@pbnjam/types'
import { ACCUMULATE_FUNCTIONS, RESULT_CODES } from '../../config'
import {
  type AccumulateHostFunctionContext,
  BaseAccumulateHostFunction,
} from './base'

/**
 * YIELD accumulation host function (Ω_♉)
 *
 * Yields accumulation trie result
 *
 * Gray Paper Specification:
 * - Function ID: 25 (yield)
 * - Gas Cost: 10
 * - Parameters: registers[7] = o
 *   - o: hash offset in memory (32 bytes)
 * - Returns: registers[7] = OK or error code
 *
 * Gray Paper Logic:
 * 1. Read hash from memory (32 bytes)
 * 2. Set the yield hash in the accumulation context
 * 3. This hash represents the result of the accumulation trie
 * 4. The hash will be used by the system to verify the accumulation result
 */
export class YieldHostFunction extends BaseAccumulateHostFunction {
  readonly functionId = ACCUMULATE_FUNCTIONS.YIELD
  readonly name = 'yield'
  readonly gasCost = 10n

  execute(context: AccumulateHostFunctionContext): HostFunctionResult {
    const { registers, ram, implications } = context
    // Extract parameters from registers
    const hashOffset = registers[7]

    // Get the current implications context
    const [imX] = implications

    // Read hash from memory (32 bytes)
    // Gray Paper pvm_invocations.tex lines 953-956:
    // h = memory[o:32] when Nrange(o,32) ⊆ readable(memory), error otherwise
    const [hashData, faultAddress] = ram.readOctets(hashOffset, 32n)
    // Gray Paper line 958: (panic, registers_7, ...) when h = error
    // Gray Paper: registers'_7 = registers_7 (unchanged) when c = panic
    if (faultAddress || !hashData) {
      // DO NOT modify registers[7] - it must remain unchanged on panic
      logger.info(`[host-calls] [${imX.id}] YIELD(${hashOffset}) <- PANIC`)
      return {
        resultCode: RESULT_CODES.PANIC,
      }
    }

    // Set the yield hash in the accumulation context
    // Gray Paper: imX.yield = h
    imX.yield = hashData

    // Set success result
    this.setAccumulateSuccess(registers)

    // Log in the requested format: [host-calls] [serviceId] YIELD(0xhash) <- OK
    const hashHex = bytesToHex(hashData)
    logger.info(`[host-calls] [${imX.id}] YIELD(${hashHex}) <- OK`)

    return {
      resultCode: null, // continue execution
    }
  }
}

File: packages/pvm-rust/src/host_functions/accumulate/yield_.rs
//! YIELD accumulation host function (Ω_♉). Gray Paper: function ID 25.
//! r7 = hash offset (o). Read 32-byte hash from memory; set yield in accumulation context; r7 = OK or panic.
//! 1:1 with AS yield.ts: read 32 bytes at o; on fault panic (r7 unchanged); else set imX.yield and OK.

use crate::config::FUNC_YIELD;
use crate::host_functions::accumulate::base::{self, codes};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};
use crate::types::bytes_to_hex;

const HASH_LEN: u32 = 32;

pub struct YieldHostFunction;

impl HostFunction for YieldHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_YIELD
    }
    fn name(&self) -> &'static str {
        "yield"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let hash_offset = context.registers[7] as u32;
        let _log_service_id = context.service_id.unwrap_or(0);

        let read_result = context.ram.read_octets(hash_offset, HASH_LEN);
        if read_result.fault_address != 0 {
            crate::host_log!(
                "[hostfn] yield PANIC: hash read fault (offset={}, fault_address={})",
                hash_offset, read_result.fault_address
            );
            crate::host_log!("[host-calls] [{}] YIELD({}) <- PANIC", _log_service_id, hash_offset);
            return HostFunctionResult::panic();
        }
        let Some(hash_data) = read_result.data else {
            crate::host_log!("[hostfn] yield PANIC: hash read returned no data");
            crate::host_log!("[host-calls] [{}] YIELD({}) <- PANIC", _log_service_id, hash_offset);
            return HostFunctionResult::panic();
        };
        if hash_data.len() != HASH_LEN as usize {
            crate::host_log!(
                "[hostfn] yield PANIC: hash length mismatch (got {}, expected {})",
                hash_data.len(),
                HASH_LEN as usize
            );
            crate::host_log!("[host-calls] [{}] YIELD({}) <- PANIC", _log_service_id, hash_offset);
            return HostFunctionResult::panic();
        }

        let _hash_hex = bytes_to_hex(&hash_data);
        match &mut context.yield_hash {
            Some(yield_hash) => **yield_hash = Some(hash_data),
            None => {
                base::set_accumulate_error(context.registers, codes::HUH);
                return HostFunctionResult::continue_execution();
            }
        }

        base::set_accumulate_success(context.registers, codes::OK);
        crate::host_log!("[host-calls] [{}] YIELD({}) <- OK", _log_service_id, _hash_hex);
        HostFunctionResult::continue_execution()
    }
}
